File: java/delete-nodes-from-linked-list-present-in-array.java

public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode prev = dummy;

  
        Set<Integer> set = new HashSet<>();

        for (int num : nums) {
            set.add(num);
        }

        while (head != null) {
            if (set.contains(head.val)) {
                prev.next = head.next;
            } else {
                prev = head;
            }
            head = head.next;
        }


        
        return dummy.next;
    }
}

File: cpp/delete-nodes-from-linked-list-present-in-array.cpp
#include <unordered_set>

#include "iostream"
#include "map"
#include "math.h"
#include "vector"

using namespace std;

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode* modifiedList(vector<int>& nums, ListNode* head) {
    unordered_set<int> s;
    ListNode* dummy = new ListNode(0);
    for (int i = 0; i < nums.size(); i++) {
      s.insert(nums[i]);
    }

    dummy->next = head;

    ListNode* prev = dummy;

    while (head) {
      if (s.find(head->val) != s.end()) {
        prev->next = head->next;
      } else {
        prev = head;
      }
      head = head->next;
    }

    return dummy->next;
  }
};

int main() {
  Solution s;

  return 0;
}